src/renderer/components/WordDragManager.tsx

```diff
@@ -0,0 +1,149 @@
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+
+import {
+  Dispatch,
+  MouseEventHandler,
+  RefObject,
+  SetStateAction,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+import useMouse, { MousePosition } from '@react-hook/mouse-position';
+import { useDispatch, useSelector } from 'react-redux';
+import { Point } from 'electron';
+import { useThrottle } from '@react-hook/throttle';
+import { dispatchOp } from '../store/undoStack/opHelpers';
+import { makeMoveWords } from '../store/undoStack/ops';
+import { selectionCleared } from '../store/selection/actions';
+import { ApplicationStore } from '../store/sharedHelpers';
+import { MouseButton, rangeLengthOne } from '../util';
+
+export type CurriedByWordIndex<T> = (wordIndex: number) => T;
+
+export type WordMouseHandler = CurriedByWordIndex<
+  (wordRef: RefObject<HTMLDivElement>) => MouseEventHandler<HTMLDivElement>
+>;
+
+export type RenderTranscription = (
+  onWordMouseDown: WordMouseHandler,
+  dragState: DragState,
+  isWordBeingDragged: CurriedByWordIndex<boolean>,
+  mouse: MousePosition,
+  mouseThrottled: MousePosition,
+  dropBeforeIndex: number | null,
+  setDropBeforeIndex: Dispatch<SetStateAction<number | null>>,
+  cancelDrag: () => void
+) => JSX.Element;
+
+export type DragState = null | {
+  offset: Point;
+  wordIndex: number;
+};
+
+interface Props {
+  renderTranscription: RenderTranscription;
+  containerRef: RefObject<HTMLDivElement>;
```

Comment:
Works well and code looks good!

Just curious about this containerRef getting prop drilled right through the entire component hierarchy. Is it possible to use `document.body` or `document.getElementBy` with a use effect (inside the DragManager) instead?

Reply:
I guess it would be possible to do that though a bit ugly - prop drilling isn’t all that much better though. Maybe react context instead? I know we already have redux but I think this is a pretty good use case for it